### PR TITLE
libi3: make resolve_tilde.c musl compatible

### PR DESCRIPTION
**Pull request by nmeum**
Since you weren't interested in providing musl compatibility yourself, this is my proposal for a more standard compatible `resolve_tilde` implementation which doesn't rely on the `GLOB_TILDE` GNU Extension.

What do you think of it?

See also: #1859

### CODE REVIEW

File: libi3/resolve_tilde.c [hunk 1]
```diff
@@ -19,27 +18,21 @@
...
-        result = scalloc(strlen(head) + (tail ? strlen(tail) : 0) + 1, 1);
-        strncpy(result, head, strlen(head));
-        if (tail)
-            strncat(result, tail, strlen(tail));
+        result = sstrndup(path, (size_t)(tilde - path));
+        result = strcat(result, home_dir);
+    }
+
+    npath = tilde + 1;
+    if (strchr(npath, '~')) {
```
Comment (nmeum) [inline libi3/resolve_tilde.c]: I was wondering if I should change that to:

```C
if (strchr(npath, '~') != NULL) {
    // …
}
```

**Comment by stapelberg** [comment]
Your implementation doesn’t cover `~user` where `user != getenv("USER")`.

How about just dropping in https://github.com/openssh/openssh-portable/blob/master/openbsd-compat/glob.c conditionally for musl?

In fact, why not add support for `GLOB_TILDE` to musl itself and thereby fix the problem for all software which uses `GLOB_TILDE`?

**Comment by Airblader** [comment]
Just FYI: looking at other projects dealing with musl, a common approach seems to be to simply not resolve tilde for musl and return the path unchanged.

**Comment by stapelberg** [comment]
Currently, I don’t think that’s a good approach. We should strive to keep all functionality in i3 available for all systems on which we run.

What’s the use-case for a musl-based system? Where does one typically use musl?

**Comment by nmeum** [comment]
> Your implementation doesn’t cover ~user where user != getenv("USER")

I could fix that but a) would you merge the patch if I would do so? (see below) b) is this really used by anyone? I mean who stores his/her i3 configuration in someone else home directory?

> How about just dropping in https://github.com/openssh/openssh-portable/blob/master/openbsd-compat/glob.c conditionally for musl?

Sounds good to me, you might want to use that on uclibc as well (see below).

> looking at other projects dealing with musl, a common approach seems to be to simply not resolve tilde for musl and return the path unchanged.

aha, that's news too me…could you name some examples? Either way I would agree with Michael on that one.

> What’s the use-case for a musl-based system? Where does one typically use musl?

Well…I use it on all my linux based computers (including my laptops and desktops), there are even some distributions which allow you to use musl instead of glibc by default these days. For instance gentoo, voidlinux or alpine.

Most software does infact compile on musl without too much patching, i3 is one of the rare cases where this isn't the case.

> In fact, why not add support for GLOB_TILDE to musl itself and thereby fix the problem for all software which uses GLOB_TILDE?

That might be a good idea, I don't know if the musl maintainers have a strong opinion regarding `GLOB_TILDE`, however, why not make i3 more portable by not relying on GNU extensions?

There might be other libc implementations out there which don't support `GLOB_TILDE` either, uclibc **might be** an additional example…

**Comment by Airblader** [comment]
Sorry, I might've been too quick. The patch I actually found was a musl patch for i3 (the irony). http://git.alpinelinux.org/cgit/aports/plain/testing/i3wm/musl.patch

Disregard my comment. I wasn't proposing this as a way forward anyway.

**Comment by stapelberg** [comment]
> That might be a good idea, I don't know if the musl maintainers have a strong opinion regarding GLOB_TILDE, however, why not make i3 more portable by not relying on GNU extensions?

GLOB_TILDE is supported by the *BSD and OS X libc, so i3 is portable across the operating systems on which it makes sense to run it. We haven’t paid too much attention to libc portability as you can tell, and I deem that far less important. After all, you can just compile i3 with a different libc.

> There might be other libc implementations out there which don't support GLOB_TILDE either, uclibc might be an additional example…

Right, but as long as nobody complains, we won’t bother with that.

Can you please suggest adding GLOB_TILDE to musl and give us the link to that communication (so that people who are interested in the progress can follow along)? Once we know what they think about it, we can figure out how to proceed here.

Side note: if you want to make the argument to include GLOB_TILDE in musl carry more weight: procps is also affected by this: https://github.com/jvvv/musl-pkg-patches/blob/master/procps_Support-libcs-without-GLOB_TILDE.patch. http://codesearch.debian.net finds 83 packages which contain the string “GLOB_TILDE”.

**Comment by Airblader** [comment]
@nmeum Any updates?

**Comment by nmeum** [comment]
I don't know what the best place is to file feature requests for musl…probably their mailing list, however, I started working on a patch for musl which adds support for `GLOB_TILDE`.

I will add a link to this comment as soon as I suggested my patch upstream…

**Pull request closed without merging**